Instant Uploads part II + more

mneuwert opened the pull request:
##  Description
Primary feature is improved instant upload enabling uploads in background as well as better control over when media is uploaded and if cellular data budget can be used for uploads.

### List of implemented improvements:

- Media upload settings (export, instant upload) moved to the separate view
- Added possibility to specify different upload folder for video instant upload
- Improved instant upload settings UI
- Added support for oC SDK feature "cellular switches". So there is an option _Media Upload_ which can be found under _Settings -> Cellular Transfers_.
- Photo album view (upload from photo library) improvements:
  * Improved album fetch and thumbnail generation performance / caching
  * Added progress indicator shown in the left top corner shown while albums are being fetched
  * Added support for iCloud based photo-stream and shared albums
  * Sorting "camera roll" smart album to the top of the list for quicker access
- Extended photo upload code to allow uploading of assets which are not stored locally and have to be downloaded from iCloud first (which is e.g. the case for shared albums).
~~- Background fetch on iOS13) using BackgroundTasks framework using combination of fetch task to check if there are new assets and import them into the sync engine.~~
~~- Background location which allows for more reliable background transfers with a trade-off that system will try to persistently persuade the user to revoke location permissions for the app.~~
~~- Please note that settings and wording used for background transfers is different on iOS13 and iOS12 or earlier.~~
~~- Optional local notifications triggered by successful background auto-upload attempts.~~

### Known issues
Since background NSURLSession is broken since iOS13 and background sessions are going to be penalised, being able to add media into import queue doesn't provide any real benefit, since the transfers can't be started, unless the app is being opened.

--> Therefore, removed background upload settings for now (commented out).

## Related Issue
#538 
#587
#711 

## Motivation and Context
see description

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## QA

Test plan: https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Version%2011.4.0/Instant%20Uploads%20II%20.md

Reports:

- [x] (1) Options are re-enabled automatically https://github.com/owncloud/ios-app/pull/714#issuecomment-648727088
- [x] (2) Target folder moved/removed/deleted https://github.com/owncloud/ios-app/pull/714#issuecomment-648734823. [MOVED to https://github.com/owncloud/ios-app/pull/662]
- [x] (3) Uploads after being notified to do in backgr https://github.com/owncloud/ios-app/pull/714#issuecomment-649481884
- [ ] (4) Path is missed after upgrading https://github.com/owncloud/ios-app/pull/714#issuecomment-651665435

Commit:
Improved various aspects of media uploads

- Changed a way how assets are exported which photo uploads significantly faster
- Refactored PHAsset+Upload.swift extensions which is now way more readable.
- Improved naming of assets
- Changed default naming of uploaded assets to the scheme legacy app was using
- Added an option allowing to preserve “original” asset names as used by iOS photo library
- Prepared code for future enhancments (exporting additional assets like RAW images and live photo video clips

Commit:
Localization of strings and other tiny changes

Commit:
Made a mechanism deriving original media asset name more reliable

Commit:
Fixed requesting image data

Commit:
Using PHAsset.creationDate as lastModifiedDate when uploading media asset

The timestamp is then set into corresponding property in the WebDAV request allowing to have proper file creation date when uploaded to oC server instance.

Commit:
Picking edited version of the photo when uploading by default

Commit:
Moved files around to have a cleaner structure in Xcode

Commit:
Using PHAssetResourceManager for export whenever possible

Commit:
Fixed small issues

Commit:
Corrected some comments

felix-schwarz committed:
Merge remote-tracking branch 'origin/milestone/1.4' into fix/uploaded_asset_naming

Commit:
Merge branch 'fix/uploaded_asset_naming' into background_media_uploads

Commit:
Some small refactoring changes

felix-schwarz committed:
Add support for cellular transfer settings
- rename StorageSettingsSection to DataSettingsSection to extend scope
- add row and cellular transfer summary to DataSettingsSection
- add new CellularSettingsViewController to present and allow control over cellular settings

felix-schwarz committed:
- Update SDK with cellular switch improvements

felix-schwarz committed:
- Update SDK to adjust concurrency budgets for sync action so that wifi-only up-/downloads can't block cellular up-/downloads entirely

Commit:
Merge remote-tracking branch 'origin/feature/cellular-switch' into feature/background_media_uploads

Commit:
Moved media upload settings into separate view

- New view for media upload settings
- Split settings into sections: Export, Auto Upload
- Changed appearance of the Upload Path setting
- Code refactoring changes

Commit:
Fixed swift lint warning

Commit:
Added cellular switch for media uploads

Commit:
Separate configuration for video auto-upload

Added posibility to select different account / path for video auto-uploads

Commit:
Fixed small issue in video upload

Commit:
Improved the UX for photo album view in upload

- Improved the way thumbnails are requested
- While view is active thumbnails for albums are cached
- Added activity indicator which is shown while albums are being fetched from photo library

Commit:
Uploading media from iCloud

- Showing cloud based photo albums in the selection UI
- Ability do upload media which is not available locally

Commit:
Improved vide exporting code

Commit:
Showing camera roll first in list of albums

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/ios-app?pullRequest=714) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/ios-app?pullRequest=714) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: Michael Neuwert<br/>:x: felix-schwarz<br/><hr/>**Michael Neuwert** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/ios-app?pullRequest=714) it.</sub>

mmattel commented:
Q: because of:
> Added possibility to specify different upload folder for video instant upload

Can you take a look on issue: https://github.com/owncloud/ios-app/issues/711
It seems that when the setting is set, you can´t change it anymore. Would be a mess if that issue gets "copied" into video upload settings...

mneuwert commented:
@mmattel yes, I am aware of #711. Will check it.

Commit:
Refactored the method for asset fetching into separate extension

hosy commented:
@mneuwert very cool!
Does it make sense to have two _Cellular Switch_ for photos and video upload?

mneuwert commented:
@hosy I just added a list of already implemented improvements.. to not forget anything.. still WIP. I already added such a switch which allows you to disable use of cellular data for photo and videos. Or are you suggesting two separate switches?

Commit:
Media upload queue refactoring

- More modular design and more maintainable code
- Started using OperationQueue for individual imports and created dedicated Operation subclass

Commit:
Forgot to add license header

Commit:
Small tweaks in media upload queue

- Changed import queue QoS to utility
- Added another cancellation point for MediaUploadOperation

Commit:
Configured main thread checker not to stop on first hit

Commit:
Background media uploads

- Improved and extended settings UI
- Using both BackgroundTasks and background location to enable background media uploads
- Local notifications informing about background uploads

hosy commented:
@mneuwert @michaelstingl I think it makes sense to decide between photo and video upload. I often take videos with a length of some minutes and I would not like to upload on the go via cellular. Because we also decide between different upload paths I would vote for two switches.

michaelstingl commented:
> not sure if it is required to have two separate switches.

Separate switches preferred.

I can imagine use cases with very long videos. (university lectures). Also other camera apps store much bigger amount of data per second. (FiLMiC Pro: 100mpbs at 4K or 50mbps at 1080)

mneuwert commented:
@michaelstingl @hosy will rename existing and add additional switch for videos

Commit:
Fixed few issues concerning auto-upload settings

Commit:
Fixed some issues and added cellular switch for videos

- Some code refactoring in PHAsset export code
- Calling the upload completion handler when placeholder item is created allowing MediaUploadQueue to proceed

Commit:
Fixed an issue with photoLibraryChangeDetected flag reset at inappropriate time

Commit:
Improved wording in background upload settings

Commit:
Refactored local notifications in separate category

Commit:
Fixed swift lint warnings

Commit:
Cleaned up code

Commit:
Improved task scheduling code

Commit:
Merge remote-tracking branch 'origin/milestone/1.4' into feature/background_media_uploads

# Conflicts:
#	ownCloud/Resources/en.lproj/InfoPlist.strings

Commit:
Re-added AVAsset extension

Commit:
Don’t show auto-upload settings if no accounts are configured

Commit:
Small improvements

- Wording for background location on iOS13 changed
- Added cell identifiers for easier creation of UI tests

Commit:
Fix of UI test was necessary since settings have moved

Commit:
Additional test for auto-upload settings added

Commit:
Fixed remaining failing UI tests

Commit:
Implemented CLLocationManagerDelegate error callback

Commit:
Merge remote-tracking branch 'origin/milestone/1.4' into feature/background_media_uploads

# Conflicts:
#	ownCloud.xcodeproj/project.pbxproj
#	ownCloud/Client/Actions/Actions+Extensions/UploadCameraMediaAction.swift
#	ownCloud/Media Uploads/MediaUploadQueue.swift
#	ownCloud/Media Uploads/PhotoKit Extensions/PHAsset+Upload.swift
#	ownCloud/Settings/MediaUploadSettingsSection.swift
#	ownCloud/Tools/URL+Extensions.swift

Commit:
Re-applied a fix lost during merging

Commit:
Changed a way location tracking is initiated

- Start when app is going into background
- Stop when app is foreground
- Switch from significant location to visit monitoring

felix-schwarz reviewed (commented):
Lots of awesome in this PR! 😍

The only real issue is the use of `OCItemTracker.swift`, because that class requests but never returns `OCCore`s on every invocation - thereby keeping the core(s) around and running indefinitely.

Fixing `OCItemTracker` should be straightforward as far as I can see (I included a suggestion in one of the comments), but what remains unclear to me is how much both the media upload code and the `Intents` depend on the `OCCore` to keep running.

For the media upload code, managing the core to keep it around for the uploads should be straightforward. As far as I can see, it should be relatively trivial to request the core from the `OCCoreManager` when uploads start - and return it when the scheduled asset list is empty.

For the `Intents` I didn't check yet.

mneuwert commented on ownCloud/Settings/BackgroundUploadsSettingsSection.swift:
```diff
@@ -0,0 +1,167 @@
+
+		if #available(iOS 13, *) {
+			footerText += " "
+			footerText += "Otherwise background media uploads using background refresh technology would depend on how often do you use the app.".localized
```
@felix-schwarz changed.

mneuwert commented on ownCloud/Settings/AutoUploadSettingsSection.swift:
```diff
@@ -0,0 +1,427 @@
+				optionSwitch.isOn = granted
+
+				if !granted {
+					let alert = UIAlertController.alertControllerForPhotoLibraryAuthorizationInSettings()
```
@felix-schwarz fixed..

mneuwert commented on ownCloud/Client/PhotoAlbumTableViewController.swift:
```diff
@@ -60,18 +61,16 @@ class PhotoAlbumTableViewController : UITableViewController, Themeable {
 
 	deinit {
 		Theme.shared.unregister(client: self)
+		PhotoAlbumTableViewCell.stopCachingThumbnails()
```
@felix-schwarz good catch.. now fixed with instantiating an instance of `PHChachingImageManager` on view controller level. So second scene won't be affected by closing first one.

mneuwert commented on ownCloud/Media Uploads/MediaUploadOperation.swift:
```diff
@@ -0,0 +1,177 @@
+		// Track the target path
+		importGroup.enter()
+
+		OCItemTracker().item(for: core.bookmark, at: path) { (_, _, item) in
```
@felix-schwarz fixed

mneuwert commented on ownCloud/Settings/AutoUploadSettingsSection.swift:
```diff
@@ -0,0 +1,427 @@
+		self.remove(rowWithIdentifier: AutoUploadSettingsSection.videoUploadBookmarkAndPathSelectionRowIdentifier)
+
+		if let bookmark = getSelectedBookmark(for: .photo), let path = userDefaults.instantPhotoUploadPath {
+			OCItemTracker().item(for: bookmark, at: path) { (error, _, pathItem) in
```
@felix-schwarz fixed `OCItemTracker`, discussed it with @hosy and both had an opinion that `OCItemTracker` should not leak `OCCore` instances since they are only used within `OCItemTracker` completion handler anyway.

felix-schwarz replied:
Thanks!

mneuwert commented on ownCloud/Settings/AutoUploadSettingsSection.swift:
```diff
@@ -0,0 +1,427 @@
+		}
+
+		if let bookmark = getSelectedBookmark(for: .video), let path = userDefaults.instantVideoUploadPath {
+			OCItemTracker().item(for: bookmark, at: path) { (error, _, pathItem) in
```
@felix-schwarz fixed as well (see above)

mneuwert commented on ownCloud/Tools/URL+Extensions.swift:
```diff
@@ -25,10 +25,14 @@ extension URL {
-	func upload(with core:OCCore?, at rootItem:OCItem, alternativeName:String? = nil, modificationDate:Date? = nil, importByCopy:Bool = false, placeholderHandler:UploadHandler? = nil, completionHandler:UploadHandler? = nil) -> Progress? {
+	func upload(with core:OCCore?, at rootItem:OCItem, alternativeName:String? = nil, modificationDate:Date? = nil, importByCopy:Bool = false, cellularSwithIdentifier:OCCellularSwitchIdentifier? = nil, placeholderHandler:UploadHandler? = nil, completionHandler:UploadHandler? = nil) -> Progress? {
 		let fileName = alternativeName != nil ? alternativeName! : self.lastPathComponent
 		var importOptions : [OCCoreOption : Any] = [OCCoreOption.importByCopying : importByCopy, OCCoreOption.automaticConflictResolutionNameStyle : OCCoreDuplicateNameStyle.bracketed.rawValue]
```
@felix-schwarz fixed

mneuwert commented on ownCloud/Media Uploads/MediaUploadQueue.swift:
```diff
@@ -23,10 +23,16 @@ import MobileCoreServices
 
 class MediaUploadQueue : OCActivitySource {
+
+	static let photoUploadCellularSwitchIdentifier = OCCellularSwitchIdentifier(rawValue: "cellular-photo-upload")
```
@felix-schwarz absolutely.. fixed

mneuwert commented on ownCloud/Media Uploads/MediaUploadQueue.swift:
```diff
@@ -43,6 +49,16 @@ class MediaUploadQueue : OCActivitySource {
 	// MARK: - Public interface
 
+	func setup() {
+		let photoCellularSwitch = OCCellularSwitch(identifier: MediaUploadQueue.photoUploadCellularSwitchIdentifier, localizedName: "Photo upload".localized, defaultValue: true, maximumTransferSize: 0)
```
@felix-schwarz fixed

mneuwert commented on ownCloud/Resources/en.lproj/InfoPlist.strings:
```diff
@@ -20,4 +20,7 @@
 "NSAppleMusicUsageDescription" = "This permission is needed for uploading media files to your server.";
 "NSPhotoLibraryAddUsageDescription" = "This permission is needed to upload photos and videos from your photo library.";
 "NSPhotoLibraryUsageDescription" = "This permission is needed to upload photos and videos from your photo library.";
+"NSLocationAlwaysAndWhenInUseUsageDescription" = "This permission is needed to enable auto-upload of photos and videos in background";
```
@felix-schwarz fixed

mneuwert commented on ownCloud/Resources/en.lproj/InfoPlist.strings:
```diff
@@ -20,4 +20,7 @@
 "NSPhotoLibraryAddUsageDescription" = "This permission is needed to upload photos and videos from your photo library.";
 "NSPhotoLibraryUsageDescription" = "This permission is needed to upload photos and videos from your photo library.";
+"NSLocationAlwaysAndWhenInUseUsageDescription" = "This permission is needed to enable auto-upload of photos and videos in background";
+"NSLocationAlwaysUsageDescription" = "This permission is needed to enable auto-upload of photos and videos in background";
```
@felix-schwarz fixed

mneuwert commented on ownCloud/Resources/en.lproj/InfoPlist.strings:
```diff
@@ -20,4 +20,7 @@
 "NSPhotoLibraryUsageDescription" = "This permission is needed to upload photos and videos from your photo library.";
+"NSLocationAlwaysAndWhenInUseUsageDescription" = "This permission is needed to enable auto-upload of photos and videos in background";
+"NSLocationAlwaysUsageDescription" = "This permission is needed to enable auto-upload of photos and videos in background";
+"NSLocationWhenInUseUsageDescription" = "This permission is needed to enable auto-upload of photos and videos in background";
```
@felix-schwarz fixed

mneuwert commented on ownCloud/Tasks/InstantMediaUploadTaskExtension.swift:
```diff
@@ -98,52 +118,8 @@ class InstantMediaUploadTaskExtension : ScheduledTaskAction {
+		return videoAssets.count
 	}
 
 	private func showFeatureDisabledAlert() {
```
@felix-schwarz looks like some left-over from the `Date.distantPast`;-) Removed

mneuwert commented on ownCloud/Tools/URL+Extensions.swift:
```diff
@@ -25,10 +25,14 @@ extension URL {
 	}
 
-	func upload(with core:OCCore?, at rootItem:OCItem, alternativeName:String? = nil, modificationDate:Date? = nil, importByCopy:Bool = false, placeholderHandler:UploadHandler? = nil, completionHandler:UploadHandler? = nil) -> Progress? {
+	func upload(with core:OCCore?, at rootItem:OCItem, alternativeName:String? = nil, modificationDate:Date? = nil, importByCopy:Bool = false, cellularSwithIdentifier:OCCellularSwitchIdentifier? = nil, placeholderHandler:UploadHandler? = nil, completionHandler:UploadHandler? = nil) -> Progress? {
```
@felix-schwarz fixed

Commit:
Fixed review findings

Commit:
Removed a WiFi requirement from pending media upload task

Commit:
Removed some unused code

Commit:
Added more logs

felix-schwarz committed:
- add experimental option to force-enable background NSURLSession usage in the app
- in code, change all instances of the usage of the term "master" to "main"

felix-schwarz committed:
Merge remote-tracking branch 'origin/milestone/1.4' into feature/background_media_uploads

Commit:
Removed an option allowing background NSURLSession

jesmrec commented:
Starting QA here...

jesmrec commented:
### (1)

1. Disable the  `Background uploads` options
2. Go back to `Settings`
3. Enter again in `Media upload`

Current: location info and upload notifications are enabled
Expected: location info and upload notifications are disabled

![Jun-24-2020 12-03-50](https://user-images.githubusercontent.com/14894746/85536011-fee04a80-b612-11ea-8402-8b5ae55482d2.gif)

iPhoneXR v13.4

jesmrec commented:
### (2)

This one is a regression that we can address to a separate issue. Just, not to forget it and get your opinion:


1. Enable AutoUploads (no matter videos or photos)
2. Set an account and target folder
3. Close the app
4. In another client or web, move/remove/rename the target folder
5. Take some content from the camera
6. Reopen the app

Current: App detects that the folder does not exist anymore, and prompts the user to decide `Retry` or `Cancel` (upload is lost in case of cancelling) for every singular pic or video. This is OK if you have taken a few photos. But, if you took 100 photos since the latest time, you will have 100 times the dialog. It is not posible to skip it.

We had a similar discussion after the first iteration of the feature, letting it be. After reworking everything, is there any other feasible solution? would it be better to address to separate issue? 


![ezgif com-optimize (6)](https://user-images.githubusercontent.com/14894746/85538480-3a7c1400-b615-11ea-8c17-7378a2252286.gif)

iPhoneXR v13.4

mneuwert commented:
### Re (2)
@jesmrec I think this is not going to be a problem when we have merged message queue feature. cc: @felix-schwarz

mneuwert commented:
### Re (1)
You just disabled background refresh.. if it is disabled, the location change would still trigger an app to check if there are new photos. May be the wording is a bit misleading. If you disable both: Auto upload of photos and videos, then the background upload options are going to be disabled.. (sort of greyed-out).

jesmrec commented:
(1)

For me this is confusing. I have three switches. But, no matters if i disable one of them, it enables by itself. Settings are there to let users decide what they want, if the system decides by itself, they stop being settings. If i disable `Background refresh` and therefore `Location info` and `Notifications` must keep enabled, i should not be able to disable it (setting it as greyed out?).

Probably i'm wrong with something. A small clarification about how these settings must behave will help the check test cases.

(2)

Let's check the same case in the message queue PR, then.

mneuwert commented:
### Re (1)
@jesmrec I agree it is indeed confusing, right now these settings more or less reflect authorisation status granted by the system which you can request but not revoke.. I would propose to add dedicated preferences which would control and e.g. disable location updates even if user has granted the app access to location information.

jesmrec commented:
> @jesmrec I agree it is indeed confusing, right now these settings more or less reflect authorisation status granted by the system which you can request but not revoke.. I would propose to add dedicated preferences which would control and e.g. disable location updates even if user has granted the app access to location information.

So, the only switch that user can control is `User background refresh`. The other two, will scape from users after being enabled. So, from scratch they can be clickable, and, when user enables them, the should keep enabled and greyed out.

mneuwert commented:
> > @jesmrec I agree it is indeed confusing, right now these settings more or less reflect authorisation status granted by the system which you can request but not revoke.. I would propose to add dedicated preferences which would control and e.g. disable location updates even if user has granted the app access to location information.
> 
> So, the only switch that user can control is `User background refresh`. The other two, will scape from users after being enabled. So, from scratch they can be clickable, and, when user enables them, the should keep enabled and greyed out (no way to disable them).

Yes, will change this..

jesmrec commented:
### (3)

This is not always reproducible, bit i notice it several times:

1. Enable auto uploads 
2. Enable all background settings
3. Switch to camera app
4. Take some photos

Current: After taking photos, notification is displayed. In the app, photos are in the correct folder but with the progress indicator in an infinite loop. By tapping on them, they are correctly opened but they do not appear in the server/webUI

Expected: Pics are uploaded in background, no needed to open the app to be uploaded

iPhoneXR v13.4

jesmrec commented:
Question :about this dialog (predetermined):

<img width="248" alt="Screenshot 2020-06-25 at 13 30 05" src="https://user-images.githubusercontent.com/14894746/85713245-7aacc680-b6e9-11ea-8ca5-6027d7205f7d.png">

if i select `Allow While Using App`, will the location services work when the app is in background or closed (so, not using the app)? or, will it be needed to enable `Always` in the device Settings? 

<img width="324" alt="Screenshot 2020-06-25 at 13 30 18" src="https://user-images.githubusercontent.com/14894746/85713456-ad56bf00-b6e9-11ea-9efb-b23575e711e7.png">

mneuwert commented:
> Question :about this dialog (predetermined):
> 
> <img alt="Screenshot 2020-06-25 at 13 30 05" width="248" src="https://user-images.githubusercontent.com/14894746/85713245-7aacc680-b6e9-11ea-8ca5-6027d7205f7d.png">
> 
> if i select `Allow While Using App`, will the location services work when the app is in background or closed (so, not using the app)? or, will it be needed to enable `Always` in the device Settings?
> 
> <img alt="Screenshot 2020-06-25 at 13 30 18" width="324" src="https://user-images.githubusercontent.com/14894746/85713456-ad56bf00-b6e9-11ea-9efb-b23575e711e7.png">

yes, while using the app means - only in foreground

mneuwert commented:
> ### (3)
> This is not always reproducible, but i notice it several times:
> 
> 1. Enable auto uploads
> 2. Enable all background settings
> 3. Switch to camera app
> 4. Take some photos
> 
> Current: After taking photos, notification is displayed. In the app, photos are in the correct folder but with the progress indicator in an infinite loop. By tapping on them, they are correctly opened but they do not appear in the server/webUI
> 
> Expected: Pics are uploaded in background, no needed to open the app to be uploaded
> 
> I will upload logs or whatever info is needed as soon as i reproduce again. I did not see the problem without the background disabled.
> 
> iPhoneXR v13.4

I think we can't get reliable background uploads since we have disabled background URL sessions in the app. I tried to enable them but turned out that Apple didn't fix them. You get penalised when start initiating background URL sessions in the app. This would work in the FileProvider but would require some IPC interface towards it to be able to tell FileProvider to upload files.

Commit:
Change the way background upload settings behave

mneuwert commented:
@jesmrec I pushed a fix for (1)

Sidenotes:
- There is no dependency actually between background location and background refresh.
- If background refresh is not active but location setting is active, upload jobs will be scheduled upon location change
- If both settings are active, upload jobs will be scheduled on both types of events
- May be the wording "Improve with location information" is a bit misleading
- Please also note, that scheduling upload jobs just means adding upload jobs to the internal upload queue and importing files by the OCCore instance in the SDK. But actual system upload tasks won't run since background NSURLSessions are broken in iOS.

jesmrec commented:
> May be the wording "Improve with location information" is a bit misleading

Since both `Background refresh` and `Location information` are independent, an option could be `Use location information`. So, users can decide which one (or both) of the options they want to use to trigger the detection.

> Please also note, that scheduling upload jobs just means adding upload jobs to the internal upload queue and importing files by the OCCore instance in the SDK. However actual system upload tasks won't run since background NSURLSessions are broken in iOS.

ok, this will not be easy to test, probably the notifications and the final status will be the same way. Will this work with the app totally closed?

mneuwert commented:
@jesmrec will check what is wrong with notifications switch.. And yes, I think it makes sense to change wording for location. Unfortunately, I had mixed impressions about the uploads.. Sometimes it looked fine, but since it is not possible to continue any NSURLSessions when App is in a background it is not going to work properly.. unless we involve file provider into file uploads.

Commit:
Removed background upload settings

Commented out and added a TODO

Commit:
Fixed some swift lint warnings

mneuwert commented:
@jesmrec I have removed background upload settings.. So, may be you can skip cellular switch testing in scope of this PR and finish the rest of QA tests. Then we can merge it into 11.4.. Then merge updated 11.4 in cellular switch branch and then you can check cellular switches for photos / videos?

jesmrec commented:
> @jesmrec I have removed background upload settings.. So, may be you can skip cellular switch testing in scope of this PR and finish the rest of QA tests. Then we can merge it into 11.4.. Then merge updated 11.4 in cellular switch branch and then you can check cellular switches for photos / videos?

yes, it's a good plan. Works for me.

jesmrec commented:
After latests discussions, i removed from test plan all the background and cellular switch test cases. The background test cases are pending of any future fix, and cellular switch cases were moved to the test plan of https://github.com/owncloud/ios-app/pull/709

jesmrec commented:
### (4)

Check this upgrade scenario out:

1. Install 1.3.2 version
2. Enable auto uploads for photos and videos, setting a correct path
3. Upgrade to the current branch
4. Open Settings > Media Upload. Auto uploads are enabled, and both kind photos and videos point to the same target folder (this is OK)
5. Change the target folder of Photo Uploads

Current: Auto upload videos get disabled
Expected: Auto upload videos keeps enabled

iPhone XR 13.4

Commit:
Fixed review finding

mneuwert commented:
@jesmrec (4) should be fixed with latest commit

jesmrec commented:
@mneuwert (4) is fixed.

I saw a little difference (not a bug) between fresh installation and upgrade from latest version:

- Fresh install: when auto uploads are disabled, corresponding path gets hidden.
- Upgrade: when auto uploads are disabled, corresponding path keeps visible.

Not a bad behaviour from my pov, we can address in separate issue. What do you think?

Commit:
Only show bookmark / path if instant upload is enabled

mneuwert commented:
@jesmrec I changed it.. now If the upload is disabled path / bookmark are not shown.

jesmrec commented:
Working fine.

From my side, this is totally approved.

felix-schwarz committed:
Merge remote-tracking branch 'origin/milestone/1.4' into feature/background_media_uploads
